opacus/grad_sample/grad_sample_module.py

```diff
@@ -425,6 +425,35 @@ def rearrange_grad_samples(
 
         return activations, backprops
 
+    def state_dict(self, *args, **kwargs) -> Dict:
+        """
+        Return the state dict of the wrapped module
+        """
+        ret_state_dict = {
+            f"_module.{key}": value
+            for key, value in self._module.state_dict(*args, **kwargs).items()
+        }
+        ret_state_dict["batch_first"] = self.batch_first
+        ret_state_dict["loss_reduction"] = self.loss_reduction
```

Comment:
I would vote against having this in the `state_dict`, especially `loss_reduction`.
They are fundamentally part of an architecture, not a state of the model. Additionally, one might want to use old weights stored in a state_dict to train with a different loss, thus changing loss reduction - and I don't think there's anything wrong with that.

`batch_first` if trickier, and I don't have a good use case, but there's always one-to-one mapping between model architecture and batch_first - and we don't store arch in the state_dict

Reply:
Fair points, both!
The way I thought of this is that these two, though not part of a model's state, are necessary to rebuild the model in the same state as before. If a new opacus user wants to save a checkpoint to be reloaded later, they might not know to save these extra variables. However, this can easily be solved with documentation and/or helpful message.

Alternatively, WDYT about saving this is as part of privacy engine checkpoints?

Reply:
>  If a new opacus user wants to save a checkpoint to be reloaded later, they might not know to save these extra variables

Do you consider loading your own checkpoint or loading one from a public hub? 

> Alternatively, WDYT about saving this is as part of privacy engine checkpoints?

I'm not sure I see how it can be loaded in a clear way? Would we override parameters set by GradSampleModule constructor for the module that was passed in?

Reply:
> Do you consider loading your own checkpoint or loading one from a public hub?

Both.

> I'm not sure I see how it can be loaded in a clear way? Would we override parameters set by GradSampleModule constructor for the module that was passed in?

I agree; it is not clear and also not neat. I do think there is value in saving metadata such as these in the checkpoint, but that would mean having to decide what to do at load time, which is not clear. 

To address this, in the current implementation, I have added support for passing your own dict to `save_checkpoint`. This dict can have all the metadata the user thinks is useful to save along with the state_dicts. At load time, we simply load the states and return the entire loaded dict (state_dicts and metadata) to the user for them to use it the way they see fit.

Reply:
> To address this, in the current implementation, I have added support for passing your own dict to save_checkpoint

Oh, that's neat, I love it